grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageDeframer.java

```diff
@@ -263,12 +263,24 @@ public boolean isStalled() {
      * @param endOfStream if {@code true}, indicates that {@code data} is the end of the stream from
      *        the remote endpoint.  End of stream should not be used in the event of a transport
      *        error, such as a stream reset.
-     * @throws IllegalStateException if {@link #close()} has been called previously or if
+     * @throws IllegalStateException if the specified {@code endOfStream} is {@code false} and
      *         this method has previously been called with {@code endOfStream=true}.
      */
     public void deframe(HttpData data, boolean endOfStream) {
         requireNonNull(data, "data");
         checkNotClosed();
+
+        // 'deframe()' could be called twice with endOfStream due to a race condition like the following:
+        //
+        // 1) HttpStreamReader.apply(), which invokes closeDeframer(), is called when
+        //    HttpResponse.whenComplete().
+        // 2) While ArmeriaClientCall requests next messages, GrpcStatus.reportStatus() is invoked if the next
```

Comment:
This doesn't seem right, why would `HttpResponse.whenComplete` have been called if there are still messages with the trailers left?

Reply:
Yes, I agreed. Let me investigate this more and try to fix the root cause.

Reply:
I've investigated more and I found the exact case.
1) HttpStreamReader received all data from publisher and add them to `unprocessed` of `deframer`.
2) gRPC client does not request the next message yet, so `deframer` still has unprocessedBytes and is not stalled
 and.
3) HttpStreamReader receives `onCompleted` signal and closes `deframer`
4) gRPC client request next messages that contain trailers so it tries to close `deframer`.